lightning/src/ln/channelmanager.rs

```diff
@@ -3242,6 +3278,29 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 		self.latest_block_height.fetch_sub(1, Ordering::AcqRel);
 		*self.last_block_hash.try_lock().expect("block_(dis)connected must not be called in parallel") = header.block_hash();
 	}
+
+	/// Blocks until ChannelManager needs to be persisted or a timeout is reached. It returns a bool
+	/// indicating whether persistence is necessary.
+	/// Note that the feature `allow_wallclock_use` must be enabled to use this function.
+	#[cfg(any(test, feature = "allow_wallclock_use"))]
+	pub fn wait_timeout(&self, max_wait: Duration) -> bool {
+		let notifier = self.total_consistency_lock.read().unwrap();
+		notifier.wait_timeout(max_wait)
+	}
+
+	/// Blocks until ChannelManager needs to be persisted.
+	pub fn wait(&self) {
+		let notifier = self.total_consistency_lock.read().unwrap();
+		notifier.wait()
```

Comment:
Hmmm...I hadn't considered this site - moving the `PersistenceNotifier` inside the `RwLock` leaves a lot of room for deadlock here - if some thread is trying to write the `ChannelManager` while another thread is blocked on `wait()` the persistence thread won't make progress as the waiting thread is (somewhat needlessly) holding the `total_consistency_lock`.

Reply:
That's technically not a deadlock, right? (checking my understanding.) In any case, it would definitely be extraneous blocking.

I agree that unnecessarily acquiring the lock unnecessarily seems more trouble than it's worth. @jkczyz I'm planning to separate them out again.

Reply:
Maybe? If a user has two threads persisted for some reason (I dunno, maybe they want notifications of persistence events multiple times?) then the thread waiting to write would end up blocking on the thread waiting to get a notification, which doesn't seem right.

Reply:
> That's technically not a deadlock, right? (checking my understanding.) In any case, it would definitely be extraneous blocking.

I don't believe it is a deadlock since waiting also involves holding a shared (read) lock.

> Maybe? If a user has two threads persisted for some reason (I dunno, maybe they want notifications of persistence events multiple times?) then the thread waiting to write would end up blocking on the thread waiting to get a notification, which doesn't seem right.

Doesn't `notify_all` prevent this? Maybe there is a race condition there. When using `BackgroundProcessor` there won't be a problem though since it uses `wait_timeout`.

Could this be solved by making `ChannelManager::wait` return the state to persist? Then the thread wouldn't need to acquire the exclusive lock. Will put some thought into this but I'm guessing this brings us back to square one since the exclusive lock is needed to get such state. There may be a design that would avoid this but probably not worth pursuing now.

Reply:
Gah. Thought I had responded already.

> I don't believe it is a deadlock since waiting also involves holding a shared (read) lock.

Correct, its not actually a deadlock, but it can lead to things waiting on each other (a writer waiting on a notification client that is waiting on something to happen) that shouldn't be waiting.

> Doesn't notify_all prevent this?

I think that would be a race - if one notification client gets run first, does something trivial, and then goes back to waiting, all before the persister makes it into the write lock, the persister will still needlessly block for some time.